File: kernels/C++/compositron/src/cdbs/fitting.cpp
#include "compositron/cdbs/fitting.hpp"
#include "compositron/core/fitting.hpp"

#include <cmath>
#include <unordered_map>


namespace compositron::cdbs::fitting {


double gauss2d(
    double x, double y, double amp, double x0, double y0,
    double sig_x, double sig_y, double phi
) {
    double cos_phi = std::cos(phi);
    double sin_phi = std::sin(phi);

    double x_rotated = (x - x0) * cos_phi - (y - y0) * sin_phi;
    double y_rotated = (x - x0) * sin_phi + (y - y0) * cos_phi;

    double u = x_rotated / sig_x;
    double v = y_rotated / sig_y;

    return amp * std::exp(-0.5 * u * u - 0.5 * v * v);
}


std::unordered_map<std::string, core::fitting::SimpleFitParam> fit_gauss2d(
    const Eigen::VectorXd& x,
    const Eigen::VectorXd& y,
    const Eigen::MatrixXd& z,
    const std::array<double, 6>& init
) {
    auto result = core::fitting::fit_generic_2d<Gauss2DResidual, 6>(x, y, z, init);

    if (result.fit_status != ceres::CONVERGENCE) {
        throw std::runtime_error("Fit failed");
    }

    std::unordered_map<std::string, core::fitting::SimpleFitParam> map;

    map["amp"] = result.opt[0];
    map["x0"] = result.opt[1];
    map["y0"] = result.opt[2];
    map["sig_x"] = result.opt[3];
    map["sig_y"] = result.opt[4];
    map["phi"] = result.opt[5];

    return map;
}


} // namespace compositron::cdbs::fitting

File: kernels/Julia/Compositron/src/CDBS/Fitting.jl
using NonlinearSolve

using ..Utils: SimpleFitParam


function gauss2d(
    x,
    y,
    amp,
    x0,
    y0,
    sig_x,
    sig_y,
    phi,
)
    cos_phi = cos(phi)
    sin_phi = sin(phi)

    dx = x - x0
    dy = y - y0

    amp * exp(
        -0.5 * ((dx * cos_phi - dy * sin_phi) / sig_x) ^ 2
        -0.5 * ((dx * sin_phi + dy * cos_phi) / sig_y) ^ 2
    )
end


function fit_gauss2d(
    x::Vector{Float64},
    y::Vector{Float64},
    z::Matrix{Float64},
    init::Vector{Float64},
)::Dict{String, SimpleFitParam}
    xx = vec([xi for _ in y, xi in x])
    yy = vec([yi for yi in y, _ in x])
    zz = vec(z)

    w = @. 1 / sqrt(max(zz, 1))

    m(u, p) = w .* (
        gauss2d.(p[1], p[2], u[1], u[2], u[3], u[4], u[5], u[6]) .- zz
    )

    prob = NonlinearLeastSquaresProblem(m, init, (xx, yy))

    sol = solve(prob, TrustRegion(; autodiff = AutoForwardDiff()))

    opt = sol.u

    Dict(
        "amp" => SimpleFitParam(opt[1], NaN),
        "x0" => SimpleFitParam(opt[2], NaN),
        "y0" => SimpleFitParam(opt[3], NaN),
        "sig_x" => SimpleFitParam(opt[4], NaN),
        "sig_y" => SimpleFitParam(opt[5], NaN),
        "phi" => SimpleFitParam(opt[6], NaN),
    )
end
